sw/airborne/modules/sensors/mag_qmc5883.c

```diff
@@ -0,0 +1,288 @@
+      // wait and do nothing
+      break;
+    case I2CTransSuccess:
+      // set to done
+      qmc.trans.status = I2CTransDone;
+      break;
+    case I2CTransFailed:
+      // set to done
+      qmc.trans.status = I2CTransDone;
+      break;
+    case I2CTransDone:
+      // do nothing
+      break;
+    default:
+      break;
+  }
+}
+
+/**
+ * QMC periodic function
+ */
+void mag_qmc5883_periodic(void)
+{
+  switch (qmc.status) {
+    case QMC5883_INIT:
+    qmc.status_debug = 0;
+      if (qmc.trans.status == I2CTransDone) {
+      // ask for i2c frame for set/reset period
+      qmc.trans.buf[0] = QMC5883_REG_RESET_PERIOD;
+      qmc.trans.buf[1] = 0x01;
+
+        if (i2c_transmit(&MAG_QMC5883_I2C_DEV, &qmc.trans, qmc.waddr, 2)){
+          // transaction OK, increment status
+          if (qmc.set_mode) {
+            qmc.status = QMC5883_SET_MODE;
+          } else {
+            qmc.status = QMC5883_REQUEST;
+          }
+          qmc.initialized = true;
+        }
+      }
+      break;
+
+    case QMC5883_SET_MODE:
+    qmc.status_debug = 1;
+      if (qmc.trans.status == I2CTransDone) {
+      // ask for i2c frame for set/reset period
+      qmc.trans.buf[0] = QMC5883_REG_CONTROL_1;
+      qmc.trans.buf[1] = (QMC5883_MODE_CONT|QMC5883_ODR_200|QMC5883_RNG_8G|QMC5883_OSR_512);
+
+        if (i2c_transmit(&MAG_QMC5883_I2C_DEV, &qmc.trans, qmc.waddr, 2)){
+          // transaction OK, increment status
+          qmc.status = QMC5883_REQUEST;
+          qmc.set_mode = false;
+        }
+      }
+      break;
+
+    case QMC5883_REQUEST:
+    qmc.status_debug = 2;
+      if (qmc.trans.status == I2CTransDone) {
+        qmc.trans.buf[0] = QMC5883_REG_DATXL; // sets register pointer to results register
+        if (i2c_transmit(&MAG_QMC5883_I2C_DEV, &qmc.trans, qmc.waddr, 1)){
+          // transaction OK, increment status
+          qmc.status = QMC5883_ACQUIRE;
+        }
+      }
+      break;
+
+    case QMC5883_ACQUIRE:
+    qmc.status_debug = 3;
+      if (qmc.trans.status == I2CTransDone) {
+        // clear buffer
+        qmc.trans.buf[0] = 0;
+        qmc.trans.buf[1] = 0;
+        if (i2c_receive(&MAG_QMC5883_I2C_DEV, &qmc.trans, qmc.raddr, 7)){
+          // transaction OK, increment status
+          qmc.status = QMC5883_PARSE;
+        }
+      }
+      break;
+
+    case QMC5883_PARSE: {
+      qmc.status_debug = 4;
+
+      // get raw data for debug
+      qmc.raw_mag_data[0] = (uint8_t)(qmc.trans.buf[0]);
+      qmc.raw_mag_data[1] = (uint8_t)(qmc.trans.buf[1]);
+      qmc.raw_mag_data[2] = (uint8_t)(qmc.trans.buf[2]);
+      qmc.raw_mag_data[3] = (uint8_t)(qmc.trans.buf[3]);
+      qmc.raw_mag_data[4] = (uint8_t)(qmc.trans.buf[4]);
+      qmc.raw_mag_data[5] = (uint8_t)(qmc.trans.buf[5]);
+      qmc.status_register.DRDY = ((uint8_t)(qmc.trans.buf[6]) >> 0) & 1;
+      qmc.status_register.OVL = ((uint8_t)(qmc.trans.buf[6]) >> 1) & 1;
+      qmc.status_register.DOR = ((uint8_t)(qmc.trans.buf[6]) >> 2) & 1;
```

Comment:
I don't think this is the right way to do this. Here, you need 3 calls of the periodic function to process a new data, so from 100Hz you go down to 33Hz. Also you use two steps to set the read register and then read the data, while it can be done in one step with `i2c_transceive` function.
So my suggestion here is:
- in the periodic function, at a merged state REQUEST_ACQUIRE, call i2c_transceive
```
qmc.trans.buf[0] = QMC5883_REG_DATXL;
i2c_transceive(&MAG_QMC5883_I2C_DEV, &qmc.trans, qmc.addr, 1, 7);
```
- in the event function, if state is PARSE and transaction is Success, parse data and send ABI event message. Don't forget to process the transaction failures or success in other sensor states.